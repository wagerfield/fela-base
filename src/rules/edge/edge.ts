import { arrayReduce, objectReduce } from "fast-loops"
import { Edge, Style } from "../../types"
import {
  isNull,
  isArray,
  isObject,
  parseEdge,
  wrapKey
} from "../../core/utils"

export interface EdgeRuleProps {
  edge: Edge
  value?: any
  prefix?: string
  postfix?: string
}

export default ({
  edge,
  value,
  prefix,
  postfix
}: EdgeRuleProps) => {
  const keys = parseEdge(edge)
  const edgeStyle: Style = {}
  // Handle null values
  if (isNull(value)) {
    return edgeStyle
  } else if (isArray(keys)) {
    // Process edge arrays
    return arrayReduce(
      keys,
      (style, key) => {
        style[wrapKey(key, prefix, postfix)] = value
        return style
      },
      edgeStyle
    )
  } else if (isObject(keys)) {
    // Process edge objects
    return objectReduce(
      keys as object,
      (style, keyValue, key) => {
        style[wrapKey(key, prefix, postfix)] = keyValue
        return style
      },
      edgeStyle
    )
  } else {
    throw new Error("invalid edge value")
  }
}
