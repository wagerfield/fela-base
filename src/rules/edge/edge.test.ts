import edge from "./edge"

describe("edge(props)", () => {
  it("throws error with invalid edge value", () => {
    expect(() => {
      edge({
        edge: JSON.parse("true")
      })
    }).toThrowErrorMatchingSnapshot()
  })

  it("renders expected snapshot when value is null", () => {
    expect(
      edge({
        edge: "left",
        value: null
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot when edge is a string", () => {
    expect(
      edge({
        edge: "left",
        value: "foo"
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot when edge is an array", () => {
    expect(
      edge({
        edge: ["left", "right"],
        value: "foo"
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot when edge is an object", () => {
    expect(
      edge({
        edge: {
          left: "foo",
          right: "bar"
        }
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot with prefixed props", () => {
    expect(
      edge({
        edge: ["left", "right"],
        value: "10px",
        prefix: "padding"
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot with postfixed props", () => {
    expect(
      edge({
        edge: ["left", "right"],
        value: "red",
        prefix: "border",
        postfix: "color"
      })
    ).toMatchSnapshot()
  })
})
