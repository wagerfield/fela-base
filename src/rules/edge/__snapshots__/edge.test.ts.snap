// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edge(props) renders expected snapshot when edge is a string 1`] = `
Object {
  "left": "foo",
}
`;

exports[`edge(props) renders expected snapshot when edge is an array 1`] = `
Object {
  "left": "foo",
  "right": "foo",
}
`;

exports[`edge(props) renders expected snapshot when edge is an object 1`] = `
Object {
  "left": "foo",
  "right": "bar",
}
`;

exports[`edge(props) renders expected snapshot when value is null 1`] = `Object {}`;

exports[`edge(props) renders expected snapshot with postfixed props 1`] = `
Object {
  "borderLeftColor": "red",
  "borderRightColor": "red",
}
`;

exports[`edge(props) renders expected snapshot with prefixed props 1`] = `
Object {
  "paddingLeft": "10px",
  "paddingRight": "10px",
}
`;

exports[`edge(props) throws error with invalid edge 1`] = `"invalid edge value"`;
